### PR TITLE
OH2-133 | Implement permission logic

### PR DESCRIPTION
**Pull request by SteveGT96**

### CODE REVIEW

File: src/components/accessories/alertBox/consts.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    ...theme.typography.caption,
+  },
+  icon: (props) => ({
+    color: theme.palette[props.type].main,
+    display: "flex",
+    alignItems: "center",
+    padding: 8,
+    height: "'100%'",
+  }),
+}));
```
Comment (axelpezzo) [inline src/components/accessories/alertBox/consts.ts:38]: We're not gonna use Styled Components as long as I remember. 

If you need to customize some styles based on props please wrape the Alert Material UI Component with a custom OH component and pass to him the props.

If you need some further customized style add a .scss file.

File: src/components/accessories/laboratory/EditLaboratoryContent.tsx [hunk 2]
```diff
@@ -88,12 +90,18 @@ export const EditLaboratoryContent: FC = () => {
...
       {open && (
-        <ExamForm
-          fields={formFields}
-          handleReset={handleReset}
-          creationMode={creationMode}
-          labWithRowsToEdit={labWithRows ?? {}}
-        />
+        <PermissionWrapper
+          permissions={
+            creationMode ? ["exam.create"] : ["exam.read", "exam.update"]
```
Comment (axelpezzo) [inline src/components/accessories/laboratory/EditLaboratoryContent.tsx:95]: Do we need the read permission in order to show the examform in edit mode?

Comment (axelpezzo) [reply src/components/accessories/laboratory/EditLaboratoryContent.tsx:95]: BEside, if we already start with the usePermission hook I think that would be great to keep the same pattern as long as possibile inside the component.

File: src/components/accessories/laboratory/Exams.tsx [hunk 3]
```diff
@@ -168,7 +172,9 @@ export const Exams: FC = () => {
       <div className="lab_labs">
         <Switch>
           <Route path={`${path}`} exact>
-            {ExamContent}
+            <PermissionWrapper permissions={["exam.read"]}>
+              {ExamContent}
+            </PermissionWrapper>
```
Comment (axelpezzo) [inline src/components/accessories/laboratory/Exams.tsx:177]: As before, keep the same pattern.

File: src/components/accessories/permissionWrapper/PermissionWrapper.tsx [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+import { IOwnProps } from "./types";
+
+export const PermissionWrapper: FC<IOwnProps> = ({
+  fallback = <CustomPermissionDenied />,
+  permissions,
+  children,
+}) => {
+  const canAccess = useAlternativePermissions(permissions);
+  return <>{canAccess === true ? children : fallback}</>;
+};
```
Comment (axelpezzo) [inline src/components/accessories/permissionWrapper/PermissionWrapper.tsx:16]: We already implemented a wrapper component called "Permission.tsx" inside libraries/permissionUtils.

Why did you need to create another one?

File: src/components/activities/newPatientActivity/NewPatientActivity.tsx [hunk 5]
```diff
@@ -108,42 +108,44 @@ const NewPatientActivity: FunctionComponent<TProps> = ({
...
+                  fields={initialFields}
+                  onSubmit={onSubmit}
+                  submitButtonLabel={t("common.submit")}
+                  resetButtonLabel={t("common.clearall")}
+                  isLoading={isLoading}
+                  shouldResetForm={shouldResetForm}
+                  resetFormCallback={resetFormCallback}
+                />
+              </PermissionWrapper>
+              <div ref={infoBoxRef}>
```
Comment (axelpezzo) [inline src/components/activities/newPatientActivity/NewPatientActivity.tsx:122]: I think that the wrapper's position is in the wrong place. I mean the patient create permission should check in three different points:
- new patient button inside the patient dashboard.
- new patient route (some control that prevent the user to access to that route)
- a check inside the form page (it's useless in due to the two previous check but I think would be great to have this too). 

The third check should wrap the entire page's UI and not only the form it's self.

File: src/components/accessories/laboratory/Exams.tsx [hunk 6]
```diff
@@ -168,7 +172,9 @@ export const Exams: FC = () => {
...
           </Route>
           <Route path={`${path}/new`}>
             <EditLaboratoryContent />
```
Comment (axelpezzo) [inline src/components/accessories/laboratory/Exams.tsx:180]: I think that even /new and /edit path sould be wrapped

File: src/components/accessories/laboratory/Exams.tsx (hunk 3 above)
Comment (axelpezzo) [inline src/components/accessories/laboratory/Exams.tsx:177]: I think that even /new and /edit path sould be wrapped

File: src/components/accessories/patientOPD/patientOPD.tsx [hunk 7]
```diff
@@ -65,6 +74,10 @@ const PatientOPD: FunctionComponent = () => {
       t("common.somethingwrong")
   ) as string;
 
+  const open = useMemo(() => {
+    return creationMode ? canCreate : canUpdate;
+  }, [creationMode, canCreate, canUpdate]);
+
```
Comment (axelpezzo) [inline src/components/accessories/patientOPD/patientOPD.tsx:80]: The "open" name is too much misleading, please rename it in something more contextual here and in the other files.

File: src/components/accessories/patientVisit/patientVisit.tsx [hunk 8]
```diff
@@ -63,6 +72,10 @@ const PatientVisit: FunctionComponent = () => {
     (state: IState) => state.patients.selectedPatient.data
   );
 
+  const open = useMemo(() => {
+    return creationMode ? canCreate : canUpdate;
+  }, [creationMode, canCreate, canUpdate]);
+
```
Comment (axelpezzo) [inline src/components/accessories/patientVisit/patientVisit.tsx:78]: As before

File: src/components/activities/editPatientActivity/EditPatientActivity.tsx [hunk 9]
```diff
@@ -128,34 +134,42 @@ const EditPatientActivity: FunctionComponent<TProps> = ({
...
+                  fields={updateFields(initialFields, patient?.data)}
+                  profilePicture={patient.data?.blobPhoto}
+                  onSubmit={onSubmit}
+                  submitButtonLabel={t("common.submit")}
+                  resetButtonLabel={t("common.reset")}
+                  isLoading={isLoading}
+                  shouldResetForm={shouldResetForm}
+                  resetFormCallback={resetFormCallback}
+                />
+              </PermissionWrapper>
```
Comment (axelpezzo) [inline src/components/activities/editPatientActivity/EditPatientActivity.tsx:148]: I think that all the jsx should be wrapped inside the permission check component. It's not wrong leave as is but I prefer that all the "protected" component (even If I cannot access to them directly as for example a modal) be stay under the wrapper. 

Please follow this here and in the other files.

File: src/components/activities/patientDetailsActivity/PatientDetailsActivity.tsx [hunk 10]
```diff
@@ -63,6 +64,7 @@ const PatientDetailsActivity: FunctionComponent<TProps> = ({
   const { id } = useParams<{ id: string }>();
   const { path, url } = useRouteMatch();
   const history = useHistory();
+  const canEdit = usePermission("patient.update");
```
Comment (axelpezzo) [inline src/components/activities/patientDetailsActivity/PatientDetailsActivity.tsx:67]: I should check if I can access to this page too (something like If I don't have any permission or if I don't have the read permission to the patient).

File: src/components/activities/patientDetailsActivity/PatientDetailsActivity.tsx [hunk 11]
```diff
@@ -375,12 +379,26 @@ const PatientDetailsActivity: FunctionComponent<TProps> = ({
                       <Route path={`${path}/admissions`}>
                         <PatientDetailsContent
                           title="Admissions"
+                          permissions={[
+                            "admission.read",
+                            "admission.create",
+                            "admission.update",
+                          ]}
```
Comment (axelpezzo) [inline src/components/activities/patientDetailsActivity/PatientDetailsActivity.tsx:386]: I think that is more simple wrap each component with a Permission checker, don't you?

File: src/components/activities/patientDetailsActivityContent/PatientDetailsActivityContent.tsx [hunk 12]
```diff
@@ -13,7 +15,13 @@ const PatientDetailsActivityContent: FunctionComponent<IProps> = ({
...
       </div>
       <div className="patientDetils__content_body">
-        <Content />
+        {permissions ? (
+          <PermissionWrapper permissions={permissions}>
+            <Content />
+          </PermissionWrapper>
+        ) : (
+          <Content />
+        )}
```
Comment (axelpezzo) [inline src/components/activities/patientDetailsActivityContent/PatientDetailsActivityContent.tsx:24]: See before.

File: src/libraries/permissionUtils/useMultiplePermissions [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+  if (!process.env.ENABLE_PERMISSIONS) {
+    return true;
+  }
+
+  return (
+    permissions.filter((permission: string) =>
+      names.includes(permission as TPermission)
+    ).length === names.length
+  );
+};
```
Comment (axelpezzo) [inline src/libraries/permissionUtils/useMultiplePermissions:20]: I'm not sure what the purpose of this.

**Comment by axelpezzo** [comment]
@SteveGT96 few notes I've noticed:

- When I try to access to a protected route and I haven't the permission to access to it, an error message should be shown on the UI (only for the route and not for each singole component as for example table or form)
- If I haven't the permission to edit and delete an element now in the table list I see an empty column, this should be handled by the Table component in order to hide it.
- If I haven't the permission to create an entity now I can't even edit his items clicking on the edit button on the table: the form still hiden.
- If i haven't the patient route permission I shoudn't be redirect to /patient when I login to the application (the same for the link on the OH logo, now it redirect me to the /patient route). I think that this could be solved by a whitelabel route (e.g. /) that check the permission and redirect me to the first route available.
- If I haven't the permission of read the patient I shouldn't be able to browser them inside the filter patient in /visit & /laboratory routes.

File: src/components/activities/visitsActivity/VisitsActivity.tsx [hunk 14]
```diff
@@ -2,12 +2,13 @@ import React, { FC } from "react";
 import { useTranslation } from "react-i18next";
 import { useSelector } from "react-redux";
 import { PATHS } from "../../../consts";
+import { Permission } from "../../../libraries/permissionUtils/Permission";
 import { TUserCredentials } from "../../../state/main/types";
 import { IState } from "../../../types";
 import AppHeader from "../../accessories/appHeader/AppHeader";
+import { CustomPermissionDenied } from "../../accessories/customPermissionDenied/CustomPermissionDenied";
```
Comment (alefalezza) [inline src/components/activities/visitsActivity/VisitsActivity.tsx:9]: Unused?

File: src/mockServer/fixtures/permissionList.js [hunk 15]
```diff
@@ -23,14 +23,18 @@ const permissionList = [
   "therapy.create",
   "therapy.update",
   "therapy.delete",
-  "visits.read",
-  "visits.create",
-  "visits.update",
-  "visits.delete",
+  "visit.read",
```
Comment (alefalezza) [inline src/mockServer/fixtures/permissionList.js:26]: Please ensure to use the same naming used in the backend

File: src/libraries/permissionUtils/usePermission.ts [hunk 16]
```diff
@@ -7,6 +7,8 @@ import { usePermissions } from "./usePermissions";
  */
 export const usePermission = (name: TPermission): boolean => {
   const permissions = usePermissions();
-
+  if (!process.env.ENABLE_PERMISSIONS) {
```
Comment (alefalezza) [inline src/libraries/permissionUtils/usePermission.ts:10]: Please declare this variable also in the .env file, in order to expose this option. I would prefer permissions to e enabled by default.

**Review by alefalezza — COMMENTED** [review]
@SteveGT96  please check my comments

**Review by alefalezza — COMMENTED** [review]
@Steve please check my comments

**Comment by SteveGT96** [comment]
> 

OK !

File: src/components/accessories/patientTriage/PatientTriage.tsx [hunk 17]
```diff
@@ -1,9 +1,18 @@
...
+  useRef,
+  useState,
+} from "react";
 import { useTranslation } from "react-i18next";
 import { useDispatch, useSelector } from "react-redux";
 import checkIcon from "../../../assets/check-icon.png";
 import { PatientExaminationDTO } from "../../../generated";
 import { updateTriageFields } from "../../../libraries/formDataHandling/functions";
+import { Permission } from "../../../libraries/permissionUtils/Permission";
+import { usePermission } from "../../../libraries/permissionUtils/usePermission";
```
[the remark this answers is not in the document]
Comment (alefalezza) [reply src/components/accessories/patientTriage/PatientTriage.tsx:15]: To remove.

File: src/components/accessories/patientVisit/patientVisit.tsx [hunk 18]
```diff
@@ -1,4 +1,10 @@
-import React, { FunctionComponent, useEffect, useRef, useState } from "react";
+import React, {
+  FunctionComponent,
+  useEffect,
+  useMemo,
```
Comment (alefalezza) [inline src/components/accessories/patientVisit/patientVisit.tsx:4]: Remove

File: src/components/accessories/patientVisit/patientVisit.tsx [hunk 19]
```diff
@@ -22,6 +28,8 @@ import { updateVisitFields } from "../../../libraries/formDataHandling/functions
 import { getWards } from "../../../state/ward/actions";
 import PatientOperation from "../patientOperation/PatientOperation";
 import { CustomDialog } from "../customDialog/CustomDialog";
+import { usePermission } from "../../../libraries/permissionUtils/usePermission";
```
[the remark this answers is not in the document]
Comment (alefalezza) [reply src/components/accessories/patientVisit/patientVisit.tsx:31]: Remove

File: src/components/accessories/permissionWrapper/PermissionWrapper.tsx [hunk 20]
```diff
@@ -0,0 +1,15 @@
+import React, { FC } from "react";
```
[the remark this answers is not in the document]
Comment (alefalezza) [reply src/components/accessories/permissionWrapper/PermissionWrapper.tsx:1]: It should be removed in favour of Permission component

**Review by alefalezza — COMMENTED** [review]
Please resolve all comments

File: src/components/accessories/patientOPD/patientOPD.tsx [hunk 21]
```diff
@@ -24,6 +24,7 @@ import { updateOpdFields } from "../../../libraries/formDataHandling/functions";
 import PatientOperation from "../patientOperation/PatientOperation";
 import { CustomDialog } from "../customDialog/CustomDialog";
 import { PatientExtraData } from "../patientExtraData/patientExtraData";
+import { Permission } from "../../../libraries/permissionUtils/Permission";
 
 const PatientOPD: FunctionComponent = () => {
   const { t } = useTranslation();
```
Comment (alefalezza) [inline src/components/accessories/patientOPD/patientOPD.tsx:30]: At line 46:
```
const [deletedObjCode, setDeletedObjCode] = useState("");
```

Is not used, can we remove it?

**Review by alefalezza — CHANGES_REQUESTED** [review]

**Pull request merged**